lib/internal/quic/core.js

```diff
@@ -0,0 +1,3341 @@
+  }
+
+  [kInspect]() {
+    const direction = this.bidirectional ? 'bidirectional' : 'unidirectional';
+    const initiated = this.serverInitiated ? 'server' : 'client';
+    const obj = {
+      id: this.#id,
+      direction,
+      initiated,
+      writableState: this._writableState,
+      readableState: this._readableState,
+      stats: {
+        dataRate: this.dataRateHistogram,
+        dataSize: this.dataSizeHistogram,
+        dataAck: this.dataAckHistogram,
+      }
+    };
+    return `QuicStream ${util.format(obj)}`;
+  }
+
+  [kTrackWriteState](stream, bytes) {
+    // TODO(@jasnell): Not yet sure what we want to do with these
+    // this.#writeQueueSize += bytes;
+    // this.#writeQueueSize += bytes;
+    // this[kHandle].chunksSentSinceLastWrite = 0;
+  }
+
+  [kUpdateTimer]() {
+    // TODO(@jasnell): Implement this later
+  }
+
+  get pending() {
+    // The id is set in the kSetHandle function
+    return this.#id === undefined;
+  }
+
+  get aborted() {
+    return this.#aborted;
+  }
+
+  get serverInitiated() {
+    return !!(this.#id & 0b01);
+  }
+
+  get clientInitiated() {
+    return !this.serverInitiated;
+  }
+
+  get unidirectional() {
+    return !!(this.#id & 0b10);
+  }
+
+  get bidirectional() {
+    return !this.unidirectional;
+  }
+
+  #writeGeneric = function(writev, data, encoding, cb) {
+    if (this.destroyed)
+      return;  // TODO(addaleax): Can this happen?
+
+    // The stream should be corked while still pending
+    // but just in case uncork
+    // was called early, defer the actual write until the
+    // ready event is emitted.
+    if (this.pending) {
+      return this.once('ready', () => {
+        this.#writeGeneric(writev, data, encoding, cb);
+      });
+    }
+
+    this[kUpdateTimer]();
+    const req = (writev) ?
+      writevGeneric(this, data, cb) :
+      writeGeneric(this, data, encoding, cb);
+
+    this[kTrackWriteState](this, req.bytes);
+  };
+
+  _write(data, encoding, cb) {
+    this.#writeGeneric(false, data, encoding, cb);
+  }
+
+  _writev(data, cb) {
+    this.#writeGeneric(true, data, '', cb);
+  }
+
+  // Called when the last chunk of data has been
+  // acknowledged by the peer and end has been
+  // called. By calling shutdown, we're telling
+  // the native side that no more data will be
+  // coming so that a fin stream packet can be
+  // sent.
+  _final(cb) {
+    // The QuicStream should be corked while pending
+    // so this shouldn't be called, but just in case
+    // the stream was prematurely uncorked, defer the
+    // operation until the ready event is emitted.
+    if (this.pending)
+      return this.once('ready', () => this._final(cb));
+
+    const handle = this[kHandle];
+    if (handle === undefined) {
+      cb();
+      return;
+    }
+
+    const req = new ShutdownWrap();
+    req.oncomplete = () => cb();
+    req.handle = handle;
+    const err = handle.shutdown(req);
+    if (err === 1)
+      return cb();
```

Comment:
The corresponding logic in `net` looks a bit different.

```js

  if (err === 1 || err === UV_ENOTCONN)  // synchronous finish
    return afterShutdown.call(req, 0);
  else if (err !== 0)
    return this.destroy(errnoException(err, 'shutdown'));

```

Reply:
not sure whether relevant, just a FYI

Reply:
Likewise, can address any issues here in subsequent PRs